### PR TITLE
check setsockopt return

### PR DESCRIPTION
**Pull request by deweerdt**

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -1706,6 +1706,14 @@ static int open_inet_listener(h2o_configurator_command_t *cmd, yoml_t *node, con
     return fd;
 }
 
+static void set_socket_buffers(int fd, unsigned sndbuf, unsigned rcvbuf)
+{
+    if (sndbuf != 0 && setsockopt(fd, SOL_SOCKET, SO_SNDBUF, &sndbuf, sizeof(sndbuf)) != 0)
+        h2o_fatal("failed to set SO_SNDBUF:%s", strerror(errno));
```
Comment (kazuho) [inline src/main.c:1712]: `set_socket_buffers` has to be MT-safe when called from `h2o_accept`. Therefore, we cannot use `strerror`. `h2o_strerror` should be used instead.

**Review by kazuho — COMMENTED** [review]
Thank you for the PR.

I'm a bit wary of crashing when we fail to set buffer sizes of a new connection, when h2o is already serving requests. At least, that's not consistent with what we do for other socket options, see `set_tcp_congestion_controller`.

If we want to be consistent with what we do for congestion controller, how about doing fprintf in `set_socket_buffers`, returning a boolean indicating success or failure? Then, we can exit if invoked during setup (i.e., from `set_quic_sockopts`), or ignore the error when invoked from `on_accept`.

Thoughts?

**Comment by gfx** [comment]
> If we want to be consistent with what we do for congestion controller, how about doing fprintf in set_socket_buffers, returning a boolean indicating success or failure? Then, we can exit if invoked during setup (i.e., from set_quic_sockopts), or ignore the error when invoked from on_accept.

I think `h2o_fatal()` is suitable here because the setsockopt fails only if its parameters are invalid, according to setsockopt(3). That is, if it fails, the code (or the configuration of h2o) is wrong. The problem is that the failure is too late -- after accepting a request. However, it is not a critical problem because most of the system checks h2o's health before making the node to accept real-world requests. 

What do you think?

**Comment by kazuho** [comment]
> However, it is not a critical problem because most of the system might (and should) check h2o's health before making the node to accept real-world requests.

I do not think that's the way small deployments are being run.

h2o running in master-worker mode relies on that property. When SIGHUP is received by the master process, the master process launches a new worker process that reads the configuration file that might be updated. Then, if the worker process does not exit immediately, the older worker process is killed. That how graceful update works.

If errors are detected sometime after the worker process is launched, then we'd see service disruption.

**Comment by gfx** [comment]
> h2o running in master-worker mode relies on that property. When SIGHUP is received by the master process, the master process launches a new worker process that reads the configuration file that might be updated. Then, if the worker process does not exit immediately, the older worker process is killed. That how graceful update works.

It makes quite sense! We should not break the mechanism of the graceful restart. Changed to use `h2o_perror() in https://github.com/h2o/h2o/pull/3095.

**Comment by gfx** [comment]
Closing in favor of https://github.com/h2o/h2o/pull/3116

**Pull request closed without merging**